# 🤖 Linear Regression: Scratch Implementation vs. Scikit-Learn

## 📝 General Project Overview
This notebook presents a custom implementation of the **Linear Regression** algorithm from scratch using **Python** and **NumPy**. 

### Core Objective:
To demonstrate the transition from the mathematical **Normal Equation** to a functional programmatic solution and validate its accuracy against the industry-standard `scikit-learn` library.

### Key Features:
* **`LinearregressionClosed` Class**: A vectorised implementation of the Ordinary Least Squares (OLS) method.
* **The Bias Trick**: Automatic handling of the intercept by augmenting the feature matrix.
* **Matrix Operations**: Leveraging the **Moore-Penrose Pseudoinverse** (`pinv`) for maximum numerical stability.
* **Benchmarking**: Performance validation on the **California Housing Dataset** using the $R^2$ metric.

---

## 📦 Phase 1: Environment Setup and Library Imports

In this initial phase, we import the essential tools for our experiment. We use **NumPy** for high-performance matrix operations and **Scikit-Learn** for data acquisition and benchmarking.

* **`numpy`**: The foundation for our mathematical operations (dot products, matrix inversion, and array manipulation).
* **`fetch_california_housing`**: A real-world dataset containing 20,640 samples with 8 features, used to predict house prices.
* **`r2_score`**: Our primary metric (Coefficient of Determination) to evaluate how well the model explains the variance in the data.
* **`LinearRegression`**: The built-in Scikit-Learn class that serves as our "Gold Standard" for accuracy comparison.

In [3]:
import numpy as np
from sklearn.datasets import fetch_california_housing
from sklearn.metrics import r2_score
from sklearn.linear_model import LinearRegression

## 🧠 Phase 2: Custom Linear Regression Class Implementation

This block contains the core logic of our **Closed-Form Solution**. By using the **Normal Equation**, we calculate the optimal weights in a single mathematical step without the need for iterative loops or learning rates.

### 📐 Mathematical Mechanics:
* **The "Bias Trick"**: We use `np.c_` to add a column of ones to the matrix $X$. This allows us to solve for the intercept ($w_0$) alongside the feature coefficients ($w_1, \dots, w_n$).
* **Normal Equation**: We solve $w = (X^T X)^{-1} X^T y$. 
* **Numerical Stability**: By using `np.linalg.pinv` (Moore-Penrose Pseudoinverse), we ensure the model works even if features are highly correlated (collinear), preventing errors when the determinant is zero.

### 🛠 Method Breakdown:
1. **`__init__`**: Initializes the model parameters (`coef_` and `intercept_`) following the naming convention of Scikit-Learn.
2. **`fit(x, y)`**: Trains the model. It augments the data, calculates the weights, and separates the first weight as the intercept.
3. **`predict(x)`**: Applies the learned weights to new data using a vectorized dot product (`@`) for maximum efficiency.

In [4]:
class LinearregressionClosed:

    def __init__(self):
        self.coef_ = None
        self.intercept_ = 0.0

    def fit(self, x, y):
        x = np.array(x)
        y = np.array(y)

        xb = np.c_[np.ones((x.shape[0], 1)), x]
        A = np.linalg.pinv(xb.T @ xb) @ xb.T @ y

        self.intercept_ = A[0]
        self.coef_ = A[1:]

    def predict(self, x):
        x = np.array(x)

        return x @ self.coef_ + self.intercept_


## 🏠 Phase 3: Data Loading and Model Training

In this phase, we apply our custom class to a real-world problem. We use the **California Housing dataset**, which is a standard benchmark for regression tasks.

### 📋 Step-by-Step Execution:
1. **Model Instantiation**: We create an instance of our `LinearregressionClosed` class.
2. **Data Acquisition**: We use `fetch_california_housing` to retrieve the features ($X$) and the target house prices ($y$).
3. **The Training Phase (`fit`)**: The model performs the matrix calculations ($X^T X$ and $\text{pinv}$) to find the optimal weights.
4. **Generating Predictions**: We use the learned weights to estimate house prices for the entire dataset.

> **Note**: Even though this dataset has 8 different features, our matrix-based implementation handles all of them simultaneously without any modifications to the code.

In [6]:
reg = LinearregressionClosed()

x, y = fetch_california_housing(return_X_y=True)

reg.fit(x, y)

y_pred = reg.predict(x)


In [7]:
print(r2_score(y, y_pred))

0.606232685199805


## 🏁 Phase 4: Final Validation and Performance Benchmarking

In this final phase, we compare our custom implementation against the industry-standard **Scikit-Learn** `LinearRegression` model. 

### 📊 Evaluation Metric: R² Score
The **Coefficient of Determination ($R^2$)** measures the proportion of variance for a dependent variable that is explained by an independent variable in a regression model.
* **1.0** indicates a perfect fit.
* **0.0** indicates the model performs no better than a simple horizontal line at the mean.

### ⚖️ The Comparison:
* **Scikit-Learn**: Uses an optimized version of the Singular Value Decomposition (SVD) to calculate weights.
* **Custom Class**: Uses our manual implementation of the Normal Equation with `pinv`.

By comparing the $R^2$ scores, we can verify if our mathematical derivation and code implementation are correct. If the scores match, it confirms that our manual approach is statistically sound and mathematically equivalent to the standard library.

In [8]:
regsk = LinearRegression()

regsk.fit(x, y)

y_pred = regsk.predict(x)

print(r2_score(y, y_pred))

0.606232685199805


## 🏁 Conclusion: The Result of Comparison

The $R^2$ scores for both our custom **`LinearregressionClosed`** and the Scikit-Learn **`LinearRegression`** are identical:

> **R² Score: 0.606232685199805**

### 🏆 Key Takeaway:
This result confirms that our manual implementation is **mathematically perfect**. By using the **Normal Equation** and the **Moore-Penrose Pseudoinverse**, we achieved the exact same level of accuracy as the industry-standard library. Our model successfully captured the underlying patterns in the California Housing dataset, proving that our "from scratch" code is robust and reliable.